interfaces/machine_parameters/ring.py

```diff
@@ -0,0 +1,588 @@
+            self.alpha_orders += [1]
+        if self.eta_orders >= 2:
+            self.alpha_orders += [2]
+        if self.eta_orders >= 3:
+            warn_message = 'The eta_orders can only be computed up to eta_2!'
+            warnings.warn(warn_message)
+
+        self.alpha_orders = np.unique(self.alpha_orders)
+
+        # Getting the momentum compaction programs in all sections
+        # and combining the programs (the missing programs are filled
+        # with zeros).
+        # The programs are reshaped to the size of the momentum program
+        for order in self.alpha_orders:
+
+            alpha_prog = []
+            alpha_name = 'alpha_%d' % (order)
+
+            for section in self.Section_list:
+
+                if hasattr(section, alpha_name):
+                    alpha_prog.append(getattr(section, alpha_name))
+                else:
+                    alpha_prog.append(ring_programs.momentum_compaction(
+                        0, order=order))
+
+            alpha_prog = ring_programs.momentum_compaction.combine_single_sections(
+                *alpha_prog, interpolation='linear')
+
+            setattr(self, alpha_name, alpha_prog.reshape(
+                self.n_sections, self.cycle_time, self.use_turns))
+
+        # Slippage factor derived from alpha, beta, gamma
+        for order in range(self.eta_orders + 1):
+            setattr(self, 'eta_%d' % (order), np.zeros(self.momentum.shape))
+        self._eta_generation()
+
+    @classmethod
+    def direct_input(cls, Particle, length, alpha_0,
+                     momentum=None, kin_energy=None, energy=None,
+                     bending_field=None, bending_radius=None, orbit_bump=None,
+                     alpha_1=None, alpha_2=None, **kwargs):
+
+        # Getting the number of sections for the length list
+        length = np.array(length, ndmin=1, dtype=float)
+        n_sections = len(length)
+
+        # Checking that at least one synchronous data input is passed
+        syncDataTypes = ('momentum', 'kin_energy', 'energy', 'B_field')
+        syncDataInput = (momentum, kin_energy, energy, bending_field)
+        assrt.single_not_none(*syncDataInput,
+                              msg='Exactly one of ' + str(syncDataTypes) +
+                              ' must be declared',
+                              exception=excpt.InputError)
+
+        # Taking the first synchronous_data input not declared as None
+        # The assertion above ensures that only one is declared
+        for func_type, synchronous_data in zip(syncDataTypes, syncDataInput):
+            if synchronous_data is not None:
+                break
+
+        # Casting synchronous data to datatype
+        if not isinstance(synchronous_data,
+                          ring_programs._synchronous_data_program):
+            synchronous_data \
+                = ring_programs._synchronous_data_program._conversions[
+                    func_type](synchronous_data)
```

Comment:
```suggestion
            synchronous_data  = getattr(ring_programs, func_type](synchronous_data)
```

Reply:
Corrected with renaming of local syncDataType